### PR TITLE
SM-5879: Add implementation to Generate Sample Inspections job

### PR DESCRIPTION
**Pull request by caoimheb**
## Description

Data - https://github.com/departmentfortransport/street-manager-data/pull/89
Jobs - https://github.com/departmentfortransport/street-manager-jobs/pull/12
Worker - https://github.com/departmentfortransport/street-manager-worker/pull/12

## Checklist

- [x] All unit tests passing
- [x] All integration tests passing
- [x] Code coverage suitable
- [x] Branch named {feature|hotfix|task}/{SM-.*}
- [x] If your pull request depends on any other, please link them
- [ ] Changes approved by your team
- [ ] Changes approved by another team
- [x] Commit messages are meaningful
- [x] I have read and understand the [Jobs release process](https://github.com/departmentfortransport/street-manager-jobs#release-process)
- [ ] Target is green https://circleci.com/gh/departmentfortransport/street-manager-jobs/tree/master
- [ ] Int UI tests passing https://circleci.com/gh/departmentfortransport/street-manager-int/tree/master
- [ ] Add `DO NOT MERGE` if you want to postpone merge

### CODE REVIEW

File: src/generate-sample-inspections/daos/sampleInspectionDao.ts [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+  ]
+
+  public constructor(@inject(TYPES.Knex) private knex: Knex) {}
+
+  public async insert(trx: Knex.Transaction, sampleInspections: SampleInspection[]): Promise<number[]> {
+    return await this.knex(this.SAMPLE_INSPECTION_TABLE).insert(sampleInspections).transacting(trx)
+  }
+
+  public async createNumberToGenerateTempTable(trx: Knex.Transaction, organisationId: number): Promise<void> {
+    return await this.knex.raw(`CREATE TEMP TABLE sample_inspection_to_generate ON COMMIT DROP AS
```
Comment (Kellym-Kainos) [inline src/generate-sample-inspections/daos/sampleInspectionDao.ts:48]: 👌

File: src/generate-sample-inspections/daos/sampleInspectionDao.ts [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+  private getCategoryBEligibleWorks(organisationId: number): Knex.QueryBuilder {
+    return this.knex.select(this.ELIGIBLE_B_WORKS_SELECT)
+    .from('work')
+    .innerJoin('sample_inspection_to_generate', 'work.promoter_organisation_id', 'sample_inspection_to_generate.promoter_organisation_id')
+    .innerJoin('site', { 'site.work_id': 'work.work_id', 'site.reinstatement_type_id ': 1 })
+    .innerJoin('reinstatement', 'reinstatement.site_id', 'site.site_id')
+    .leftJoin('sample_inspection', { 'sample_inspection.work_id': 'work.work_id', 'sample_inspection.inspection_category_id': 2 })
+    .leftJoin('reinspection', 'work.work_id', 'reinspection.work_id')
+    .whereRaw(`reinstatement.reinstatement_date >= NOW() - INTERVAL '6 MONTHS'`)
```
Comment (Kellym-Kainos) [inline src/generate-sample-inspections/daos/sampleInspectionDao.ts:129]: does this cover null dates

File: src/generate-sample-inspections/daos/sampleInspectionDao.ts [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+  private getCategoryCEligibleWorks(organisationId: number): Knex.QueryBuilder {
+    return this.knex.select(this.ELIGIBLE_C_WORKS_SELECT)
+    .from('work')
+    .innerJoin('sample_inspection_to_generate', 'work.promoter_organisation_id', 'sample_inspection_to_generate.promoter_organisation_id')
+    .innerJoin('site', { 'site.work_id': 'work.work_id', 'site.reinstatement_type_id ': 1 })
+    .innerJoin('reinstatement', 'reinstatement.site_id', 'site.site_id')
+    .leftJoin('sample_inspection', { 'sample_inspection.work_id': 'work.work_id', 'sample_inspection.inspection_category_id': 3 })
+    .leftJoin('reinspection', 'work.work_id', 'reinspection.work_id')
+    .whereRaw('reinstatement.end_date >= NOW()')
+    .whereRaw(`reinstatement.end_date <= NOW() + INTERVAL '3 MONTHS'`)
```
Comment (Kellym-Kainos) [inline src/generate-sample-inspections/daos/sampleInspectionDao.ts:146]: same about null dates

File: src/generate-sample-inspections/daos/sampleInspectionDao.ts (hunk 2 above)
Comment (caoimheb) [reply src/generate-sample-inspections/daos/sampleInspectionDao.ts:129]: `reinstatement_date` is not nullable on the `reinstatement` table, so shouldn't be any nulls

File: src/generate-sample-inspections/daos/sampleInspectionDao.ts (hunk 3 above)
Comment (caoimheb) [reply src/generate-sample-inspections/daos/sampleInspectionDao.ts:146]: `end_date` is not nullable on the `reinstatement` table, so shouldn't be any nulls

File: src/generate-sample-inspections/services/generateSampleInspectionsService.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    this.logger.log(`Starting sample inspection generation for job id: ${jobId}`)
+
+    const trx: Knex.Transaction = await super.startTransaction()
+
+    try {
+      const organisationId = await this.asyncJobService.getJobOrganisation(jobId)
+
+      await this.sampleInspectionDao.createNumberToGenerateTempTable(trx, organisationId)
+
+      const categoryA: SampleInspection[] = await this.sampleInspectionDao.generateCategoryASampleInspections(trx, organisationId)
```
[the remark this answers is not in the document]
Comment (caoimheb) [reply src/generate-sample-inspections/services/generateSampleInspectionsService.ts:33]: 👌

File: src/generate-sample-inspections/services/generateSampleInspectionsService.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+  public async generateSampleInspections(jobId: number): Promise<void> {
+    this.logger.log(`Starting sample inspection generation for job id: ${jobId}`)
+
+    const trx: Knex.Transaction = await super.startTransaction()
+
+    try {
+      const organisationId = await this.asyncJobService.getJobOrganisation(jobId)
+
+      await this.sampleInspectionDao.createNumberToGenerateTempTable(trx, organisationId)
```
Comment (cpaul92) [inline src/generate-sample-inspections/services/generateSampleInspectionsService.ts:30]: 🙌

File: src/generate-sample-inspections/services/generateSampleInspectionsService.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    super(knex)
+  }
+
+  public async generateSampleInspections(jobId: number): Promise<void> {
+    this.logger.log(`Starting sample inspection generation for job id: ${jobId}`)
+
+    const trx: Knex.Transaction = await super.startTransaction()
+
+    try {
+      const organisationId = await this.asyncJobService.getJobOrganisation(jobId)
```
Comment (cpaul92) [inline src/generate-sample-inspections/services/generateSampleInspectionsService.ts:28]:
```suggestion
      const organisationId: number = await this.asyncJobService.getJobOrganisation(jobId)
```

**Pull request merged**